### ISSUE TITLE
Green highlight for variables referencing a Node

### ISSUE THREAD

**Issue by RPicster** [godotengine/godot-proposals#8208]
### Describe the project you are working on

Dome Keeper, Engraving, Jam Games, Prototypes and others - This request comes mostly out of discussions about code conventions at Bippinbits.

### Describe the problem or limitation you are having in your project

At Bippinbits we had a big discussion about code conventions. I personally use the "officially" recommended GDScript style for my projects but at work projects we use a different style.

For our next project I pushed more towards using the official style (we used camelCase so far instead of snake_case).
Now we had a bigger discussion about code readability.

The biggest problem / hinderance with using the recommended `snake_case` for all variables comes up when using the classic:
```GDScript
@onready var player = $Player
```

I like to use this when things are running very often (e.g. _process or a loop) vs directly using `$Player`
Now there is a problem with readability.

Using `$Player` will give a nice, green highlight which helps a ton with recognizing that there is a node being referenced.
Using `player` (from the onready var) does not have this highlight and could as well be a variable storing whatever.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I would love to have the green highlights for variables/constants storing a Node reference. This would help to make code easier to parse and to recognize node references instantly.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is a sample snippet with the different styles:

master  - using $ (or %) to access nodes gives a nice green color which makes node references very easy to see.
![godot_green_highlight_proposal_using_dollar](https://github.com/godotengine/godot-proposals/assets/9423774/1cca3234-bd78-4ef5-a9ad-d53786b085bf)

Storing a reference to a node via `@onready` is recommended in many cases - but this leaves us with "gray" code which is much harder to read - and to make out differences between other variables and the node references.
![godot_green_highlight_proposal_using_onready](https://github.com/godotengine/godot-proposals/assets/9423774/a1cb15f9-796b-458f-becc-873eadf3ba21)

This is an example how an imlpementation could look.
![godot_green_highlight_proposal_new](https://github.com/godotengine/godot-proposals/assets/9423774/2752aa0d-ee3c-4e9c-9e17-ee36d1631d9f)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not that I am aware of.

### Is there a reason why this should be core and not an add-on in the asset library?

I don't think there is a way to make this happen outside of core.

**Comment by Mickeon** [comment]
I don't really like the default color for this case, but part of me also sees the appeal of it. I too sometimes use the shorthand because it makes it easier to tell when the node is interacting with its children, at a glance.

**Comment by GuyUnger** [comment]
The idea is interesting, too many colors can make readability worse but since the green is already used for node references and accessing nodes is so core to godot it could be helpful.

In your example you're using an untyped variable, you could get weird things like:
```
@onready var player = $Player

func _ready():
    player = true
    player # this boolean is green?
```

If its instead for any typed variable that references a node, would it also work for parameters?
```
func button_pressed(button: Button): # button parameter is green?
    button # green?
```

**Comment by samsface** [comment]
Would this apply to any typed Node? e.g., `var node:Node = $SomeNode`
Or only nodes assigned during `@onready`?

**Comment by RPicster** [comment]
Thinking about it, depending it on the type could make even more sense. Then I think it could be as "simple" as a setting to highlight all references to classes extending "Node".

**Comment by dalexeev** [comment]
I would prefer to solve this with naming rather than highlighting. Highlighting identifiers in different colors depending on the data type looks too colorful for me. Also this will not work in other editors, GitHub, documentation, etc.

For example, if you have slider and label Volume, then the following naming is logical:

```gdscript
var _volume: int = 100
@onready var _volume_label: Label = %VolumeLabel
@onready var _volume_slider: HSlider = %VolumeSlider
```

**Comment by SlugFiller** [comment]
While using it untyped is indeed undesirable, I think it should apply based on usage, not type. i.e. only apply on:
```gdscript
@onready var ... = $...
```
As in, both `@onready` and `$` must be present.

As for, "what if it is modified later", it would be nice if there was `@onready const` which would forbid manual modification. This would make this a guarantee.

@dalexeev Just from your own example, notice that `_volume` and `_volume_label` cannot be separated at a glance (same naming style). And yet, the desire is for the latter to be different from the prior. As in, for `_volume_label` to be highlighted, while `_volume` is not.

To use naming for this, you'd need a much more unique, obvious, and hence, ugly, solution. e.g. `__NODE__volume_label`. This is not a desirable solution.

**Comment by YuriSizov** [comment]
Syntax highlight reflects what each token in the code is, not what it holds. Breaking that expectation is not desirable, because it complicates understanding the written script, not simplifies it. Also, as @dalexeev pointed out, nothing else does this, this wouldn't translate to any other programming language or any other code editor out there.

Using descriptive names is your best option. There should also be tooltips on hover that describe variables as far as the static analyzer can understand them.

**Comment by dalexeev** [comment]
> Using descriptive names is your best option. There should also be tooltips on hover that describe variables as far as the static analyzer can understand them.

* See #1393.

**Comment by RPicster** [comment]
@dalexeev @YuriSizov I agree that highlighting every data type *differently* would make it very chaotic.
But and **optional** highlighting (could be easily solved through theming) of all Node extending types could be very helpful to make it easier to read - make out a difference when it is about a node vs a "normal" variable.

**Comment by dalexeev** [comment]
This is difficult to implement without false positives, given how the highlighter is currently implemented. And besides, this will require the highlighter to use analyzer (at the moment the highlighter does not even use parser). If there is an error in the script, this may stop working.

Personally, I prefer it when tokens of the same type are highlighted equally. Or at least identifiers that are highlighted in different colors have noticeably different behavior, and are not just highlighted on a "would be nice" basis. `Node` variables are not different from other variables.

We have a few exceptions to this rule, but I don't think that further increasing the features will be wise, people can offer more of them, each proposal will have some reasonable justification, but in the end the code will look too colorful.

1. (already done) built-in types vs native classes vs global scripts, global functions;
2. (partially done) members vs locals;
3. (suggested here) `Node` variables of a separate color;
4. function parameters, because it is convenient and reasonable, and also someone does not like the `p_` prefix;
5. static members, because this is also convenient and reasonable;
6. private members, because Kate does it;
7. unused variables, because JetBrains IDEs do it;
8. and so on.

**Comment by SlugFiller** [comment]
> this will require the highlighter to use analyzer 

Won't the parser be enough? If you look for the pattern `@ready var ... = $...`, this can easily be found at the AST level. An optional extra step is to also tell apart shadowing variables or parameters, but this can also be done from the AST. However, it might actually be better to not do it, for the same reason it's not done for built-ins and globals - It makes shadowing visually look different than other variable definitions, implying that you shouldn't be doing it that way. Given that shadowing is generally discouraged in GDScript, this behavior is good.

> We have a few exceptions to this rule, but I don't think that further increasing the features will be wise, people can offer more of them, each proposal will have some reasonable justification, but in the end the code will look too colorful.

I know you intended what follows as a counter-example, but you've mostly made me kind of want several of them. Rainbow-colored code? Yes please. Everything can always be made configurable, so it fits with each user's individual palettes. And palates.

**Comment by dalexeev** [comment]
> Won't the parser be enough? If you look for the pattern `@ready var ... = $...`, this can easily be found at the AST level.

I meant the later suggestion:

> Would this apply to any typed Node? e.g., `var node:Node = $SomeNode`

And even if we limit ourselves to `@onready var ... = $...`, this will still require the parser that is not currently used. Although the highlighter can also break (for example if you don't close a quote), it is more resistant to syntax errors than the parser, which in some cases cannot recover from panic mode, which generates a huge list of cascading errors.

**Comment by RPicster** [comment]
> A lot of good conversation so far.
> 
> Just good to re-read the contributor best practices document sometimes, and see how the proposal compares to each point. https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html
> 
> Reading over the guidelines document, this proposal scores really high on points 1-3. On 1 & 2 ) This is a very real problem that is annoying devs today. On 3) This is not a complex problem but is a VERY frequent problem. How many of our scripts start with assigning a bunch of nodes to variables? Like... all of them. It's annoying and error prone.
> 
> **So, forgetting about the solution proposed, I think we've identified a real problem here. It's kinda dumb we have to manually cache node variables.**
> 
> I'd propose an alternative approach that targets the real problem, let's make $ lookups better? So something like:
> 
> ```gdscript
> @onready cache($Somenode) # under the hood, calls get_node("Somenode") and assigns to a hidden variable.
> 
> func foo():
>    $Somenode.bar() # this now uses the hidden variable and avoid the cost of calling get_node.
> ```
> 
> Though having said all that. I don't think it'd do any harm to allow syntax highlighting per type. I get there's a lot of people saying "They'd prefer" something else. That's fine. Let's just make it optional and off by default and organizations can do whatever they want with it.

Very interesting. I truly like using `$` and `%` beecause the way it highlights in the editor.
So if I can just do that without having to worry about the cost, it would be a great option.

**Comment by SlugFiller** [comment]
There's already a couple of proposals about caching `$` and `%` variables some how. Most have been rejected on account of `@onready var` doing the job. But looking at this proposal, it's made more clear that it is not.

This might be worth raising in the next GDScript meeting. Both this proposals, and all the other ones. Once we recognize that a problem exists, we can consider what's the cleanest way to deal with it.
